pintc/src/asm_gen/asm_builder.rs

```diff
@@ -200,45 +168,66 @@ impl<'a> AsmBuilder<'a> {
-
-                let access_mem_idx = self.storage_expr_to_mem_idx[expr];
+                        ))
+                    },
+                    |ty| {
+                        Ok((
+                            ty.storage_keys(handler, contract)? as i64,
+                            ty.size(handler, contract)? as i64,
+                        ))
+                    },
+                )?;
 
-                // Read the storage keys into memory
                 asm.extend([
+                    // Allocate enough memory for KREX or KRNG
+                    PUSH(access_size + 2 * num_keys),
+                    ALOC,
+                    PUSH(Self::KEY_RANGE_MEM_IDX_STACK_LOC),
+                    STOS,
+                    // Read the storage keys into memory
                     PUSH(num_keys),
-                    PUSH(access_mem_idx), // mem idx for the output of KREX or KRNG
+                    PUSH(Self::KEY_RANGE_MEM_IDX_STACK_LOC),
+                    LODS,
                     if is_extern { KREX } else { KRNG },
-                ]);
-
-                // Now read the actual data
-                asm.extend([
-                    PUSH(access_mem_idx + num_keys * 2), // where the actual data lives
-                    PUSH(expr_size - 1),                 // size of the data (excluding the tag)
+                    // Read the actual data
+                    PUSH(Self::KEY_RANGE_MEM_IDX_STACK_LOC),
+                    LODS,
+                    PUSH(num_keys * 2),
+                    ADD,                 // where the actual datalives
+                    PUSH(expr_size - 1), // size of the data (excluding the tag)
                     LODR,
                 ]);
 
                 // Sum the sizes of all the values read. These are laid out as follows in
                 // memory:
                 // `[a_addr, a_len, b_addr, b_len, a_value, b_value, ...]`
                 if num_keys > 1 {
-                    asm.push(PUSH(0));
-                    asm.extend(
-                        (0..num_keys).flat_map(|i| [PUSH(access_mem_idx + 2 * i + 1), LOD, ADD]),
-                    );
+                    asm.extend([
+                        PUSH(Self::KEY_RANGE_MEM_IDX_STACK_LOC),
+                        LODS,
+                        PUSH(num_keys * 2),
+                        LODR,
+                        PUSH(0),
+                    ]);
+                    asm.extend((0..num_keys).flat_map(|_| [ADD, SWAP, POP]));
                 } else {
-                    asm.extend([PUSH(access_mem_idx + 1), LOD]);
+                    asm.extend([
+                        PUSH(Self::KEY_RANGE_MEM_IDX_STACK_LOC),
+                        LODS,
+                        PUSH(1),
+                        ADD,
+                        LOD,
+                    ]);
                 }
 
+                // We can now free the memory used by KRNG OR KREX
+                asm.extend([PUSH(Self::KEY_RANGE_MEM_IDX_STACK_LOC), LODS, FREE]);
+
```

Comment:
Do we need to do this?  I'm kinda thinking our model can be nodes just allocate as much as they want and we have a single `FREE` at the end.  I guess we could exhaust memory on a big loop which allocates?  But once we have the compute op everything will be made up of tiny nodes and so mid-execution `FREE` really won't be necessary IMO.

Reply:
Yeah you're right.. no need to do this really. And memory for constraint nodes doesn't need to be clear at all.. it just gets dropped by the VM eventually anyways.